space-provider/kcp/kcp.go

```diff
@@ -0,0 +1,244 @@
+/*
+Copyright 2023 The KubeStellar Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package providerkcp
+
+import (
+	"context"
+	"strings"
+	"sync"
+
+	"github.com/go-logr/logr"
+
+	k8sapierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/tools/clientcmd"
+	api "k8s.io/client-go/tools/clientcmd/api"
+	"k8s.io/klog/v2"
+
+	kcpv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/core/v1alpha1"
+	tenancyv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/tenancy/v1alpha1"
+	tenancyclient "github.com/kcp-dev/kcp/pkg/client/clientset/versioned/typed/tenancy/v1alpha1"
+
+	clusterprovider "github.com/kubestellar/kubestellar/pkg/space-manager/providerclient"
+)
+
+// KcpClusterProvider is a cluster provider for KCP workspaces.
+type KcpClusterProvider struct {
+	ctx        context.Context
+	logger     logr.Logger
+	kcpConfig  string
+	clientset  tenancyclient.WorkspaceInterface
+	workspaces map[string]string
+	watch      clusterprovider.Watcher
+}
+
+// New returns a new KcpClusterProvider
+func New(kcpConfig string) (*KcpClusterProvider, error) {
+	ctx := context.Background()
+	logger := klog.FromContext(ctx)
+	config, err := clientcmd.RESTConfigFromKubeConfig([]byte(kcpConfig))
+	if err != nil {
+		return nil, err
+	}
+	clientset, err := tenancyclient.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+	c := &KcpClusterProvider{
+		ctx:        ctx,
+		logger:     logger,
+		kcpConfig:  kcpConfig,
+		clientset:  clientset.Workspaces(),
+		workspaces: make(map[string]string),
+	}
+	return c, nil
+}
+
+func (k *KcpClusterProvider) Create(name string, opts clusterprovider.Options) error {
+	k.logger.V(2).Info("Creating KCP workspace", "name", name)
+	ws := &tenancyv1alpha1.Workspace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+		Spec: tenancyv1alpha1.WorkspaceSpec{},
+	}
+	_, err := k.clientset.Create(k.ctx, ws, metav1.CreateOptions{})
+	if err != nil && !k8sapierrors.IsAlreadyExists(err) {
+		return err
+	}
+	k.logger.V(2).Info("Created KCP workspace for space", "name", name)
+	return nil
+}
+
+func (k *KcpClusterProvider) Delete(name string, opts clusterprovider.Options) error {
+	k.logger.V(2).Info("Deleting KCP workspace", "name", name)
+	return k.clientset.Delete(k.ctx, name, metav1.DeleteOptions{})
+}
+
+// ListSpacesNames is N/A for KCP
```

Comment:
I think it's redundant for the other providers as well?

Reply:
I agree. Let's remove this func from the interface in separate PR.